### DISCUSSION TITLE
interface selection between two border routers

### DISCUSSION THREAD

**Discussion by timmydlt** [openthread/openthread#9174] [Q&A] [answered]
Topology :
“End Device A” -----thread-----“Border Router A” --- ETHERNET--- “Border Router B”------thread------ “End Device B”
                                                                                --- thread---

1. If two border routers are connected through both thread and ethernet, is there a way to force traffic passing through only ethernet? 
2. If the thread connection is lost (out of range) but two border routers are still ethernet connected, could end device A and B communicate with each other? If so, what are the necessary settings on the border router?

**Comment by jwhui** [comment]
It sounds like you are looking for the Thread-over-Infrastructure feature, which can be enabled with the [OTBR_TREL](https://github.com/openthread/ot-br-posix/blob/7c81d3d96cc0eeccb57c76e995a1e5d9fbcf104e/etc/cmake/options.cmake#L106) option. The discussion in https://github.com/openthread/openthread/discussions/8478 provides some good context.

> If two border routers are connected through both thread and ethernet, is there a way to force traffic passing through only ethernet?

If both Thread and Ethernet are available, TREL will generally prefer Ethernet. That said, TREL will continuously evaluate the link quality across both Thread and Ethernet and select the one that provides better connectivity.

> If the thread connection is lost (out of range) but two border routers are still ethernet connected, could end device A and B communicate with each other? If so, what are the necessary settings on the border router?

Yes. That said, it is also possible to achieve this specific use case without TREL. The two end devices will be in different Thread partitions, with their own IPv6 ULA prefixes. The end devices need to discover each other's routable IPv6 addresses and use that.

**Reply by timmydlt** [reply]
We have a test case setting ot macfilter on border router A to deny border router B, then ot ping border router A's onlink address from end device B and the ping failed. 

Is there a setup guide for border routers so that  end devices need could discover each other's routable IPv6 addresses?

**Reply by jwhui** [reply]
> We have a test case setting ot macfilter on border router A to deny border router B, then ot ping border router A's onlink address from end device B and the ping failed.

If by "onlink" you mean mesh-local, then what you observe is expected. Mesh-local addresses are, by definition, local to a given Thread mesh topology and cannot be routed outside the Thread mesh network.

> Is there a setup guide for border routers so that end devices need could discover each other's routable IPv6 addresses?

We do have this codelab [Thread Border Router - Bidirectional IPv6 Connectivity and DNS-Based Service Discovery](https://openthread.io/codelabs/openthread-border-router), which provides an introduction to DNS-based Service Discovery.

**Reply by timmydlt** [reply]
> > We have a test case setting ot macfilter on border router A to deny border router B, then ot ping border router A's onlink address from end device B and the ping failed.
> 
> If by "onlink" you mean mesh-local, then what you observe is expected. Mesh-local addresses are, by definition, local to a given Thread mesh topology and cannot be routed outside the Thread mesh network.
> 
I ping the ip with prefix given by RA to ip network (fd12:3456:7812:5678:22fe:eeff:fe11:f471  in this test).

br-lan    Link encap:Ethernet  HWaddr 20:FE:EE:11:F4:71
          inet addr:192.168.11.5  Bcast:192.168.11.255  Mask:255.255.255.0
          inet6 addr: 2000::2/64 Scope:Global
          inet6 addr: fe80::22fe:eeff:fe11:f471/64 Scope:Link
          inet6 addr: fe80::ffff/64 Scope:Link
          inet6 addr: fd12:3456:7812:5678:22fe:eeff:fe11:f471/64 Scope:Global

**Reply by jwhui** [reply]
Can you provide more information about your setup, including:
1. Git commit id(s)
2. Hardware platform(s)
3. Build steps

Can you provide the output of the following OT CLI command:
```
netdata show
```

Can you provide the output of the following Linux command:
```
ip -6 route show
```

**Reply by timmydlt** [reply]
> Can you provide more information about your setup, including:
> 
>     1. Git commit id(s)
> 
ot-br-posix:　ee57634fd3ccb2566cca71afee0f9a6b11735be9
ot-br-posix -  openthread: 242c7cc7c3a80c985476689b64cdb6d8fbcf5b6d
rcp（siliconlab） :  OpenThread 2.2.1.0 (GitHub-91fa1f455)

>     2. Hardware platform(s)
> 
Siliconlabs MG21(RCP) + ARM SoC Linux 
>     3. Build steps
> 
> 
CMAKE_OPTIONS+= \
       -DBUILD_TESTING=OFF \
       -DCMAKE_INSTALL_PREFIX=/usr \
       -DOTBR_BORDER_AGENT=ON \
       -DOTBR_BORDER_ROUTING=ON \
       -DOTBR_INFRA_IF_NAME=\"br-lan\" \
       -DOTBR_MDNS="avahi" \
       -DOTBR_OPENWRT=ON \
       -DOTBR_SRP_ADVERTISING_PROXY=ON \
       -DOTBR_NAT64=ON \
       -DOT_FIREWALL=ON \
       -DOT_POSIX_SETTINGS_PATH=\"/etc/openthread\" \
       -DOT_READLINE=OFF \
       -DOT_MULTIPAN_RCP=ON \
       -DOT_RCP_RESTORATION_MAX_COUNT=10 \
       -DOT_POSIX_CONFIG_RCP_BUS=CPC \
       -DOT_CHANNEL_MANAGER=OFF \
       -DOT_MESH_DIAG=ON \
       -DOT_CHANNEL_MONITOR=OFF \
       -DOT_DNSSD_SERVER=ON

> Can you provide the output of the following OT CLI command:
> 
> ```
> netdata show
> ```
border router A
ot-ctl netdata show
Prefixes:
fd42:c0ad:f5ef:1::/64 paos low bc00
Routes:
fd42:c0ad:f5ef:2:0:0::/96 sn low bc00
fd12:3456:7812:5678::/64 s med bc00
fd6c:e52f:e3e8:1::/64 s med bc00
Services:
44970 5d fddead00beef00009aec9cef6d17ff1cd11f s bc00
Done

border router B
ot-ctl netdata show
Prefixes:
fd6c:e52f:e3e8:1::/64 paos low 8000
Routes:
fd12:3456:7812:5678::/64 s med 8000
fd6c:e52f:e3e8:2:0:0::/96 sn low 8000
fd42:c0ad:f5ef:1::/64 s med 8000
Services:
44970 5d fddead00beef00004b81c180078c91a6d11f s 8000
Done
> 
> Can you provide the output of the following Linux command:
> 
> ```
> ip -6 route show
> ```
border router A
ip -6 route show
2000::/64 dev br-lan proto kernel metric 256 pref medium
fd12:3456:7812:5678::/64 dev br-lan proto kernel metric 256 expires 1711sec pref medium
fd42:c0ad:f5ef:1::/64 dev wpan0 proto kernel metric 256 pref medium
fdde:ad00:beef::/64 dev wpan0 proto kernel metric 256 pref medium
fe80::/64 dev br-lan proto kernel metric 256 pref medium
fe80::/64 dev ath0 proto kernel metric 256 pref medium
fe80::/64 dev ath2 proto kernel metric 256 pref medium
fe80::/64 dev wpan0 proto kernel metric 256 pref medium
default via 2000::1 dev br-lan proto static metric 1024 pref medium

border router B
ip -6 route show
2000::/64 dev br-lan proto kernel metric 256 pref medium
fd12:3456:7812:5678::/64 dev br-lan proto kernel metric 256 expires 1777sec pref medium
fd6c:e52f:e3e8:1::/64 dev wpan0 proto kernel metric 256 pref medium
fdde:ad00:beef::/64 dev wpan0 proto kernel metric 256 pref medium
fe80::/64 dev br-lan proto kernel metric 256 pref medium
fe80::/64 dev ath0 proto kernel metric 256 pref medium
fe80::/64 dev ath2 proto kernel metric 256 pref medium
fe80::/64 dev wpan0 proto kernel metric 256 pref medium
default via 2000::1 dev br-lan proto static metric 1024 pref medium

**Reply by jwhui** [reply, marked as answer]
The linux routing tables do not show routes to each BR.
- BRA should show a route to `fd6c:e52f:e3e8:1::/64`
- BRB should show a route to `fd42:c0ad:f5ef:1::/64`

Each BR should be transmitting ICMPv6 Router Advertisement (RA) messages with a Route Information Option (RIO) advertising the prefix to their respective Thread networks. Upon receiving these RA messages, the Linux kernel should process the RIO options and install a route towards the other BR. RIO processing does require the proper [sysctl settings](https://github.com/openthread/ot-br-posix/blob/852fc69e72354156deb6d4962d1cd808fbedd51d/script/_border_routing#L45).
